service/storage_service.cc

```diff
@@ -2623,8 +2630,8 @@ future<> storage_service::on_change(gms::inet_address endpoint, const gms::appli
     // of its removal, after handle_state_normal has already been called for
     // the new one. Without the check, the do_update_system_peers_table call
     // overwrites the IP back to its old value.
-    // In essence, the code under the 'if' should fire if the given IP is a normal_token_owner.
-    if (ep && *ep == endpoint && tm.is_normal_token_owner(host_id)) {
+    // In essence, the code under the 'if' should fire if the given IP is a normal node.
```

Comment:
Do I understand correctly -- this is how we differ from Cassandra, which doesn't store zero-token nodes in system.peers, and we do want to store them, so we don't lose them after restart in RECOVERY mode?

Reply:
(and we don't lose the IP mappings)

Reply:
Yes, but now I'm not sure if this change is fine. It's needed for zero-token nodes, but maybe it incorrectly affects other token owners with no tokens at the moment.

Reply:
I added the "state == normal" check. Now this code should be equivalent to the old code, but additionally work for zero-token nodes in the normal state.

Reply:
Obsolete comments, there were a lot of changes in the topology-related logic in newer versions of this PR. I'm resolving it.